alacritty_terminal/src/input.rs

```diff
@@ -587,40 +605,62 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
 
         let button_changed = self.ctx.mouse().last_button != button;
 
+        let report_modes =
+            TermMode::MOUSE_REPORT_CLICK | TermMode::MOUSE_DRAG | TermMode::MOUSE_MOTION;
+        let report_mode_intersection = self.ctx.terminal().mode().intersects(report_modes);
+        let side = self.ctx.mouse().cell_side;
+
         self.ctx.mouse_mut().click_state = match self.ctx.mouse().click_state {
             ClickState::Click
                 if !button_changed && elapsed < self.mouse_config.double_click.threshold =>
             {
                 self.ctx.mouse_mut().block_url_launcher = true;
-                self.on_mouse_double_click(button, point);
+                if self.is_expand_selection(button, modifiers, report_mode_intersection) {
+                    if let Some(point) = point {
+                        self.ctx.update_selection_as(point, side, SelectionType::Semantic);
+                    }
+                } else {
+                    self.on_mouse_double_click(button, point);
+                }
                 ClickState::DoubleClick
             }
             ClickState::DoubleClick
                 if !button_changed && elapsed < self.mouse_config.triple_click.threshold =>
             {
                 self.ctx.mouse_mut().block_url_launcher = true;
-                self.on_mouse_triple_click(button, point);
+                if self.is_expand_selection(button, modifiers, report_mode_intersection) {
+                    if let Some(point) = point {
+                        self.ctx.update_selection_as(point, side, SelectionType::Lines);
+                    }
+                } else {
+                    self.on_mouse_triple_click(button, point);
+                }
                 ClickState::TripleClick
             }
             _ => {
                 // Don't launch URLs if this click cleared the selection
                 self.ctx.mouse_mut().block_url_launcher = !self.ctx.selection_is_empty();
 
-                self.ctx.clear_selection();
-
-                // Start new empty selection
-                let side = self.ctx.mouse().cell_side;
-                if let Some(point) = point {
-                    if modifiers.ctrl {
-                        self.ctx.block_selection(point, side);
-                    } else {
-                        self.ctx.simple_selection(point, side);
+                if self.is_expand_selection(button, modifiers, report_mode_intersection) {
+                    // update current selection
+                    if let Some(point) = point {
+                        self.ctx.update_selection_as(point, side, SelectionType::Simple);
+                    }
+                } else {
+                    self.ctx.clear_selection();
+
+                    // Start new empty selection
+                    let side = self.ctx.mouse().cell_side;
+                    if let Some(point) = point {
```

Comment:
We unpack the point from the option in both of these branches. Seems like it's better extracted to outside of them so the code isn't unnecessarily duplicated.